File: backend/utils/emailService.js
/**
 * Email Service Utility
 * =======================
 * Sends transactional emails for booking events.
 * Uses nodemailer with Gmail SMTP (configure in .env).
 * All methods are async and called non-blocking (fire-and-forget) from controllers.
 *
 * To enable: set EMAIL_USER and EMAIL_PASS in .env
 * Gmail: generate an App Password at https://myaccount.google.com/apppasswords
 */

const nodemailer = require('nodemailer');

// Create reusable transporter
const createTransporter = () =>
  nodemailer.createTransport({
    host: process.env.EMAIL_HOST || 'smtp.gmail.com',
    port: parseInt(process.env.EMAIL_PORT || '587'),
    secure: false,
    auth: {
      user: process.env.EMAIL_USER,
      pass: process.env.EMAIL_PASS,
    },
  });

// Check if email is configured
const isEmailConfigured = () =>
  !!(process.env.EMAIL_USER && process.env.EMAIL_PASS);

// ── Shared HTML template wrapper ─────────────────────────────────────────────
const htmlTemplate = (title, body) => `
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8"/>
  <style>
    body { font-family: Arial, sans-serif; background: #f4f4f4; margin: 0; padding: 20px; }
    .card { background: white; border-radius: 8px; padding: 24px; max-width: 500px; margin: auto; }
    h2   { color: #1a73e8; }
    .detail { background: #f8f9fa; padding: 12px; border-radius: 6px; margin: 12px 0; }
    .footer { margin-top: 20px; font-size: 12px; color: #888; }
    .badge-confirmed { color: #1e7e34; font-weight: bold; }
    .badge-pending   { color: #856404; font-weight: bold; }
    .badge-cancelled { color: #721c24; font-weight: bold; }
    .badge-rejected  { color: #721c24; font-weight: bold; }
  </style>
</head>
<body>
  <div class="card">
    <h2>Campus Facility Booking System</h2>
    <h3>${title}</h3>
    ${body}
    <div class="footer">
      This is an automated message. Please do not reply to this email.<br/>
      &copy; ${new Date().getFullYear()} Campus Booking System
    </div>
  </div>
</body>
</html>
`;

// ── Email methods ─────────────────────────────────────────────────────────────

/**
 * Send booking confirmation / submission email.
 */
const sendBookingConfirmation = async ({ to, userName, facilityName, date, start_time, end_time, status, bookingId }) => {
  if (!isEmailConfigured()) return;

  const statusLabel = status === 'confirmed'
    ? '<span class="badge-confirmed">✅ Confirmed</span>'
    : '<span class="badge-pending">⏳ Pending Approval</span>';

  const transporter = createTransporter();
  await transporter.sendMail({
    from:    process.env.EMAIL_FROM || process.env.EMAIL_USER,
    to,
    subject: `Booking #${bookingId} – ${status === 'confirmed' ? 'Confirmed' : 'Submitted'} | ${facilityName}`,
    html: htmlTemplate(
      status === 'confirmed' ? 'Booking Confirmed!' : 'Booking Submitted',
      `
      <p>Hi ${userName},</p>
      <p>Your booking has been ${status === 'confirmed' ? '<strong>confirmed</strong>' : 'submitted and is awaiting admin approval'}.</p>
      <div class="detail">
        <strong>Booking ID:</strong> #${bookingId}<br/>
        <strong>Facility:</strong>   ${facilityName}<br/>
        <strong>Date:</strong>       ${date}<br/>
        <strong>Time:</strong>       ${start_time} – ${end_time}<br/>
        <strong>Status:</strong>     ${statusLabel}
      </div>
      `
    ),
  });
};

/**
 * Send status-change notification (approved / rejected / completed).
 */
const sendStatusUpdate = async ({ to, userName, facilityName, status, admin_notes, bookingId }) => {
  if (!isEmailConfigured()) return;

  const badge = `<span class="badge-${status}">${status.toUpperCase()}</span>`;
  const transporter = createTransporter();
  await transporter.sendMail({
    from:    process.env.EMAIL_FROM || process.env.EMAIL_USER,
    to,
    subject: `Booking #${bookingId} Status Updated: ${status.toUpperCase()} | ${facilityName}`,
    html: htmlTemplate(
      `Booking ${status.charAt(0).toUpperCase() + status.slice(1)}`,
      `
      <p>Hi ${userName},</p>
      <p>Your booking #${bookingId} for <strong>${facilityName}</strong> has been updated.</p>
      <div class="detail">
        <strong>New Status:</strong> ${badge}<br/>
        ${admin_notes ? `<strong>Admin Notes:</strong> ${admin_notes}` : ''}
      </div>
      `
    ),
  });
};

/**
 * Send cancellation notice.
 */
const sendCancellationNotice = async ({ to, userName, facilityName, date, start_time, bookingId }) => {
  if (!isEmailConfigured()) return;

  const transporter = createTransporter();
  await transporter.sendMail({
    from:    process.env.EMAIL_FROM || process.env.EMAIL_USER,
    to,
    subject: `Booking #${bookingId} Cancelled | ${facilityName}`,
    html: htmlTemplate(
      'Booking Cancelled',
      `
      <p>Hi ${userName},</p>
      <p>Your booking has been <strong>cancelled</strong>.</p>
      <div class="detail">
        <strong>Booking ID:</strong> #${bookingId}<br/>
        <strong>Facility:</strong>   ${facilityName}<br/>
        <strong>Date:</strong>       ${date}<br/>
        <strong>Start:</strong>      ${start_time}<br/>
        <strong>Status:</strong>     <span class="badge-cancelled">CANCELLED</span>
      </div>
      `
    ),
  });
};

module.exports = { sendBookingConfirmation, sendStatusUpdate, sendCancellationNotice };

File: backend/src/main/java/com/campus/booking/service/EmailService.java
package com.campus.booking.service;

import com.campus.booking.entity.Booking;
import jakarta.mail.MessagingException;
import jakarta.mail.internet.MimeMessage;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.time.format.DateTimeFormatter;

@Service
@Slf4j
public class EmailService {

    private final JavaMailSender mailSender;

    @Value("${app.mail.from}")
    private String fromAddress;

    @Value("${app.mail.enabled:false}")
    private boolean enabled;

    private static final DateTimeFormatter DATE_FMT = DateTimeFormatter.ofPattern("MMMM d, yyyy");
    private static final DateTimeFormatter TIME_FMT = DateTimeFormatter.ofPattern("HH:mm");

    public EmailService(JavaMailSender mailSender) {
        this.mailSender = mailSender;
    }

    @Async
    public void sendBookingConfirmation(Booking booking) {
        if (!enabled) return;

        String statusLabel  = booking.getStatus().name().equals("CONFIRMED")
                ? "Confirmed" : "Submitted (Pending Approval)";
        String subject = String.format("Booking #%d – %s | %s",
                booking.getId(), statusLabel, booking.getFacility().getName());

        String html = buildHtml(
                booking.getUser().getName(),
                subject,
                booking.getFacility().getName(),
                booking.getDate().format(DATE_FMT),
                booking.getStartTime().format(TIME_FMT),
                booking.getEndTime().format(TIME_FMT),
                booking.getStatus().name(),
                null,
                booking.getId()
        );

        send(booking.getUser().getEmail(), subject, html);
    }

    @Async
    public void sendStatusUpdate(Booking booking) {
        if (!enabled) return;

        String subject = String.format("Booking #%d Status Updated: %s | %s",
                booking.getId(), booking.getStatus().name(), booking.getFacility().getName());

        String html = buildHtml(
                booking.getUser().getName(),
                subject,
                booking.getFacility().getName(),
                booking.getDate().format(DATE_FMT),
                booking.getStartTime().format(TIME_FMT),
                booking.getEndTime().format(TIME_FMT),
                booking.getStatus().name(),
                booking.getAdminNotes(),
                booking.getId()
        );

        send(booking.getUser().getEmail(), subject, html);
    }

    @Async
    public void sendCancellationNotice(Booking booking) {
        if (!enabled) return;

        String subject = String.format("Booking #%d Cancelled | %s",
                booking.getId(), booking.getFacility().getName());

        String html = buildHtml(
                booking.getUser().getName(),
                subject,
                booking.getFacility().getName(),
                booking.getDate().format(DATE_FMT),
                booking.getStartTime().format(TIME_FMT),
                booking.getEndTime().format(TIME_FMT),
                "CANCELLED",
                null,
                booking.getId()
        );

        send(booking.getUser().getEmail(), subject, html);
    }

    // ── Internals ─────────────────────────────────────────────────────

    private void send(String to, String subject, String html) {
        try {
            MimeMessage msg = mailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(msg, true, "UTF-8");
            helper.setFrom(fromAddress);
            helper.setTo(to);
            helper.setSubject(subject);
            helper.setText(html, true);
            mailSender.send(msg);
            log.debug("Email sent to {}: {}", to, subject);
        } catch (MessagingException e) {
            log.warn("Failed to send email to {}: {}", to, e.getMessage());
        }
    }

    private String buildHtml(String userName, String title, String facilityName,
                               String date, String startTime, String endTime,
                               String status, String adminNotes, Long bookingId) {

        String badgeColor = switch (status) {
            case "CONFIRMED" -> "#16a34a";
            case "PENDING"   -> "#d97706";
            case "CANCELLED" -> "#dc2626";
            case "REJECTED"  -> "#7c3aed";
            default          -> "#6b7280";
        };

        String notesHtml = adminNotes != null && !adminNotes.isBlank()
                ? "<p><strong>Admin notes:</strong> " + adminNotes + "</p>" : "";

        return """
            <!DOCTYPE html>
            <html>
            <head><meta charset="UTF-8"></head>
            <body style="font-family:sans-serif;background:#f3f4f6;margin:0;padding:24px">
              <div style="max-width:560px;margin:auto;background:#fff;border-radius:8px;
                          padding:32px;box-shadow:0 2px 8px rgba(0,0,0,.08)">
                <h2 style="margin-top:0;color:#1e3a8a">Campus Booking System</h2>
                <p>Hi %s,</p>
                <div style="background:#f9fafb;border-radius:6px;padding:16px;margin:16px 0">
                  <p style="margin:0 0 8px"><strong>Facility:</strong> %s</p>
                  <p style="margin:0 0 8px"><strong>Date:</strong> %s</p>
                  <p style="margin:0 0 8px"><strong>Time:</strong> %s – %s</p>
                  <p style="margin:0"><strong>Status:</strong>
                    <span style="background:%s;color:#fff;padding:2px 10px;
                                 border-radius:12px;font-size:12px">%s</span>
                  </p>
                </div>
                %s
                <p style="color:#6b7280;font-size:12px">Booking #%d</p>
              </div>
            </body>
            </html>
            """.formatted(userName, facilityName, date, startTime, endTime,
                    badgeColor, status, notesHtml, bookingId);
    }
}
